cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java

```diff
@@ -250,4 +263,116 @@ private static String getFailMessage(@NotNull Context ctx,
         return "Request for: " + ctx.req.getRequestURI() + " denied. Missing roles: " + missing;
     }
 
+
+    /**
+     * Return create an ApiKey for future authentication and authorization.
+     * @param p Principal object, to get the username
+     * @param keyName Friendly name for this key
+     * @param expires when this key expires; can be null to never expire
+     * @return The created ApiKey
+     */
+    public ApiKey createApiKey(DataApiPrincipal p, ApiKey sourceData) throws CwmsAuthException {
+        
+        try {
+            if(!p.getName().equalsIgnoreCase(sourceData.getUserId())) {
+                throw new CwmsAuthException(ONLY_OWN_KEY_MESSAGE, HttpCode.UNAUTHORIZED.getStatus());
+            }
+            SecureRandom randomSource = SecureRandom.getInstanceStrong();
+            String key = randomSource.ints(48,122) // allow a-zA-Z0-9
```

Comment:
Personally, I'd probably use something like
`randomSource.ints((char)'0',(char)'z')`
to make it clearer to the reader what the range, and exclusions are.